File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

void decision(int amount_digits, int firstdigit, int seconddigit, int security);

int main(void)
{
    long number;
    do
    {
        number = get_long("Credit Card Number: ");
    }
    while (number < 1 || number > 9999999999999999);
    
    // Get amount of digits
    int amount_digits = floor(log10(number)) + 1;
    
    // Initialize variables
    int firstdigit;
    int seconddigit;
    long temp_number = number;
    int temp;
    int sumeven = 0;
    int sumodd = 0;
    
    for (int i = 1; i < amount_digits + 1; i++)
    {
        temp = temp_number % 10;
        temp_number = (temp_number - temp) / 10;
        
        // Check if first or second digit from front
        if (i == amount_digits)
        {
            firstdigit = temp;
        }
        else if (i == amount_digits - 1)
        {
            seconddigit = temp;
        }
        
        // Check if every other number
        if (i % 2 == 0)
        {
            temp = temp * 2;
            if (temp < 10)
            {
                sumeven = sumeven + temp;
            }
            else
            {
                sumeven = sumeven + temp % 10 + (temp -  temp % 10) / 10;
            }
        }
        // Sum of odd numbers
        else 
        {
            sumodd = sumodd + temp;
        }
    }
    
    // Finding security number
    int security = (sumodd + sumeven) % 10;
    
    // Final Decision
    decision(amount_digits, firstdigit, seconddigit, security);
    
}

void decision(int amount_digits, int firstdigit, int seconddigit, int security)
{
    // Check if AMEX
    if (amount_digits == 15 && security == 0 && firstdigit == 3)
    {
        if (seconddigit == 4 || seconddigit == 7)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    //Check if Master
    else if (amount_digits == 16 && security == 0 && firstdigit == 5)
    {
        if (seconddigit == 1 || seconddigit == 2 || seconddigit == 3 || seconddigit == 4 || seconddigit == 5)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    // Check if Visa
    else if ((amount_digits == 13 || amount_digits == 16) && security == 0 && firstdigit == 4)
    {
        printf("VISA\n");
    }
    // Otherwise invalid
    else
    {
        printf("INVALID\n");
    }
}

File: Week 6/pset6/credit/credit.py
from cs50 import get_string

number = get_string("Enter credit card number: ")

if number.isdecimal() and (len(number) == 13 or len(number) == 15 or len(number) == 16):
    odd = 0
    even = 0
    for i in range(1, len(number) + 1):
        # even numbers
        if i % 2 == 0:
            if int(number[-i]) * 2 >= 10:
                lastnum = (int(number[-i]) * 2) % 10
                firstnum = (int(number[-i]) * 2 - lastnum) / 10
                even = even + lastnum + firstnum
            else:
                even = even + int(number[-i]) * 2
        # odd numbers
        else:
            odd = odd + int(number[-i])

    # security number
    security = int(odd + even)
    
    # real card
    if (security % 10 == 0):
        if len(number) == 15 and number[0] == "3" and (number[1] == "4" or number[1] == "7"):
            print("AMEX")
        elif len(number) == 16 and number[0] == "5" and (number[1] == "1" or number[1] == "2" or number[1] == "3" or number[1] == "4" or number[1] == "5"):
            print("MASTERCARD")
        elif len(number) == 16 and number[0] == "4":
            print("VISA")
        else:
            print("INVALID")
        
    # fake card
    else:
        print("INVALID")
    
else:
    print("INVALID")